src/Nethermind/Nethermind.Merge.Plugin/BlockProduction/PayloadPreparationService.cs

```diff
@@ -88,56 +89,165 @@ protected virtual Block ProduceEmptyBlock(string payloadId, BlockHeader parentHe
-                    ImproveBlock(payloadId, parentHeader, payloadAttributes, newBestBlock, startDateTime, blockImprovementContext.BlockFees);
+                    timeRemainingInSlot = minDelay;
                 }
-                else
+
+                // progress = 0  => at the very start of the slot
+                // progress = 1  => at the very end of the slot
+                double progress = timeUsedInSlot.TotalSeconds / _timePerSlot.TotalSeconds;
+
+                // Clamp progress to [0, 1] just in case:
+                progress = Math.Clamp(progress, 0.0, 1.0);
+
+                // We use a cubic curve (progress^3) so that improvement builds 
+                // start off quite spaced out (less frequent at the start) and 
+                // rapidly become more frequent as we near the end of the slot.
+                progress *= progress * progress;
+
+                // We'll interpolate between two fractional rates:
+                // - fractionStart (1/6): a slower build rate at the start
+                // - fractionEnd   (1/480): a faster build rate near the end
+                const double fractionStart = 1.0 / 6.0;
+                const double fractionEnd = 1.0 / 480.0;
+                // Slot Timeline: 0% -------------------------- 100%
+                //                |        (long gap)         | 
+                //    [Block Improvement #1]        <--- big delay here
+                // 
+                //                                   (medium gap)
+                //                                       [Block Improvement #2]
+                //                                           (small gap)
+                //                                              [Block Improvement #3]
+                //                                                (tiny gap)
+                //                                                   [Block Improvement #4]
+                //
+                double currentFraction = fractionStart + (fractionEnd - fractionStart) * progress;
+                // Dynamic delay = currentFraction * (remaining slot time)
+                // So near the start: delay is bigger (slower improvement)
+                // Near the end: delay shrinks, allowing more frequent improvements
+                dynamicDelay = TimeSpan.FromSeconds(timeRemainingInSlot.TotalSeconds * currentFraction);
+            }
+            else
+            {
+                // In tests, we override the dynamic strategy with a fixed delay
+                dynamicDelay = _improvementDelay.Value;
+            }
+
+            // Ensure we don't exceed the allowed time window by checking if
+            // we can finish the next improvement build before 1.3x slot time.
+            DateTimeOffset whenWeCouldFinishNextProduction = DateTimeOffset.UtcNow + dynamicDelay + lastBuildTime;
+
+            if (whenWeCouldFinishNextProduction > slotPlusThirdFinishTime)
```

Comment:
You know that in practice you need to publish a block in first 3rd of the slot?

Reply:
Clarified in comment; cancellation stops the building, so is mostly a break in case never cancelled

Reply:
Also the first 3rd of slot should be a full slot since last first 3rd of slot